x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/pivot/CompositeBucketsChangeCollector.java

```diff
@@ -486,26 +672,42 @@ public QueryBuilder buildFilterQuery(long lastCheckpointTimestamp, long nextchec
     }
 
     @Override
-    public boolean processSearchResponse(final SearchResponse searchResponse) {
+    public Map<String, Object> processSearchResponse(final SearchResponse searchResponse) {
         final Aggregations aggregations = searchResponse.getAggregations();
         if (aggregations == null) {
-            return true;
+            return null;
         }
 
-        final CompositeAggregation agg = aggregations.get(compositeAggregation.getName());
-
-        Collection<? extends Bucket> buckets = agg.getBuckets();
-        afterKey = agg.afterKey();
+        // every collector reports if the collection of changes is done after processing, if all are done, the indexer
+        // will not run a query for change collection again
+        boolean isDone = true;
 
-        if (buckets.isEmpty()) {
-            return true;
+        // collect changes from aggregations added by field collectors
+        for (FieldCollector fieldCollector : fieldCollectors.values()) {
+            isDone &= fieldCollector.collectChangesFromAggregations(aggregations);
         }
 
-        for (FieldCollector fieldCollector : fieldCollectors.values()) {
-            fieldCollector.collectChanges(buckets);
+        // collect changes from composite buckets
+        final CompositeAggregation compositeAggResults = aggregations.get(COMPOSITE_AGGREGATION_NAME);
+
+        // xor: either both must exist or not exist
+        assert (compositeAggregation == null ^ compositeAggResults == null) == false;
```

Comment:
[optional] Alternatively, you could express it this way:
```
        assert (compositeAggregation == null) == (compositeAggResults == null);
```
but I'm not sure which one is more readable.